### PR TITLE
TIMRLINKN-55 Export Groups and their Users into Database

### PR DESCRIPTION
**Pull request by gugmaster**

### CODE REVIEW

File: timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+                {
+                    description = "Nice animation series kids",
+                    externalId = "77C8",
+                    name = "Paw Patrol",
+                    parentExternalId = "77C1"
+                };
+                
+                var user1 = new User()
+                {
+                    uuid = "9d33c475-0da2-4b21-95b9-feca948cc80a"
```
Comment (daberni) [inline timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs:51]: could be single liners for better readability

File: timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+                var group1 = memoryContext.Group.Single(g => g.ExternalId == "99C4");
+                
+                Assert.AreEqual("Series with airplanes for kids", group1.Description);
+                Assert.AreEqual("Super Wings", group1.Name);
+                Assert.AreEqual("77C1", group1.ParentalExternalId);
+            }
+            
+            {
+                var memoryContext = new DatabaseContext(options);
+                var group1 = memoryContext.Group.Single(g => g.ExternalId == "77C8");
```
Comment (daberni) [inline timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs:106]: group1 + group1

why are you doing this in extra scopes each time? 🤔

File: timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            logger = loggerFactory.CreateLogger<ProjectTimeDatabaseExportAction>();
+            this.context = context;
+            this.groupService = groupService;
+        }
+
+        public async Task Execute()
+        {
+            // We don't migrate when in memory database or SQLite Databases are used, otherwise unit tests would fail.
+            if (context.Database.IsRelational() && context.Database.GetDbConnection().GetType() != typeof(Microsoft.Data.Sqlite.SqliteConnection))
```
Comment (daberni) [inline timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs:26]: Can we deduplicate this logic?

File: timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            {
+                var pendingMigrations = (await context.Database.GetPendingMigrationsAsync()).ToList();
+                if (pendingMigrations.Any())
+                {
+                    logger.LogInformation($"Running Database Migration... ({string.Join(", ", pendingMigrations)})");
+                    await context.Database.MigrateAsync();
+                }
+            } 
+
+            await context.Database.EnsureCreatedAsync();
```
Comment (daberni) [inline timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs:36]: EnsureCreatedAsync additionally to migration?

File: timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+                context.Update(databaseGroup);
+                await context.SaveChangesAsync();
+
+                logger.LogInformation($"Created or updated Group with Name: {databaseGroup.Name}");
+                
+                groupDictionary.Remove(databaseGroup.Id);
+                
+                var groupUsers = await groupService.GetGroupUsers(group);
+                var groupUsersDatabase = new List<GroupUsers>();
```
Comment (daberni) [inline timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs:65]: whats this for? why not add it directly?

File: timrlink.net.CLI/DatabaseContext.cs [hunk 6]
```diff
@@ -61,8 +72,9 @@ public class ProjectTime
     {
         [Key]
         public Guid UUID { get; set; }
-
         public string User { get; set; }
+
```
Comment (daberni) [inline timrlink.net.CLI/DatabaseContext.cs:76]: what are these blank lines?

File: timrlink.net.CLI/Migrations/20221125103413_AddUserGroupsMappingTable.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace timrlink.net.CLI.Migrations
+{
+    public partial class AddUserGroupsMappingTable : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.AddColumn<Guid>(
+                name: "UserUUID",
```
Comment (daberni) [inline timrlink.net.CLI/Migrations/20221125103413_AddUserGroupsMappingTable.cs:11]: is this the same as in TIMRLINK-53? would this result in conflicts?

File: timrlink.net.CLI/config.json [hunk 8]
```diff
@@ -1,7 +1,9 @@
 {
   "timrSync": {
-    "identifier": "",
-    "token": ""
+    "identifier": "gugmaster",
+    "token": "09cd770cffa34db69c65d3ec8fb48572",
```
Comment (daberni) [inline timrlink.net.CLI/config.json:4]: please don't commit credentials

File: timrlink.net.Core/Service/GroupService.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            this.logger = logger;
+            this.timrSync = timrSync;
+        }
+
+        public async Task<IList<Group>> GetGroups()
+        {
+            var getGroupsResponse = await timrSync.GetGroupsAsync(new GetGroupsRequest1(new API.GetGroupsRequest())).ConfigureAwait(false);
+            
+            var groups = getGroupsResponse.GetGroupsResponse1;
+            logger.LogDebug($"Total user count: {groups.Length}");
```
Comment (daberni) [inline timrlink.net.Core/Service/GroupService.cs:27]: user?

File: timrlink.net.Core/ServiceFactory.cs [hunk 10]
```diff
@@ -86,5 +86,10 @@ public static IWorkItemService BuildWorkItemService(ILoggerFactory loggerFactory
         {
             return new WorkItemService(loggerFactory.CreateLogger<WorkItemService>(), timrSync);
         }
+        
+        public static IGroupService BuildGroupService(ILoggerFactory loggerFactory, TimrSync timrSync)
```
Comment (daberni) [inline timrlink.net.Core/ServiceFactory.cs:90]: I am a bit confused, why did we need this?

**Review by daberni — DISMISSED** [review]

File: timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs (hunk 2 above)
Comment (gugmaster) [reply timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs:106]: This one is from database. The first one is `API.Group`. Do you think I should rename it?

File: timrlink.net.CLI/config.json (hunk 8 above)
Comment (gugmaster) [reply timrlink.net.CLI/config.json:4]: 😱

File: timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            logger = loggerFactory.CreateLogger<ProjectTimeDatabaseExportAction>();
+            this.context = context;
+            this.groupService = groupService;
+        }
+
+        public async Task Execute()
+        {
+            // We don't migrate when in memory database or SQLite Databases are used, otherwise unit tests would fail.
+            if (context.Database.GetDbConnection().GetType() != typeof(Microsoft.Data.Sqlite.SqliteConnection))
```
Comment (daberni) [inline timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs:26]: Can we deduplicate this migration logic?

File: timrlink.net.CLI/timrlink.net.CLI.csproj [hunk 12]
```diff
@@ -24,6 +24,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <PackageReference Include="Microsoft.Data.Sqlite" Version="5.0.17" />
```
Comment (daberni) [inline timrlink.net.CLI/timrlink.net.CLI.csproj:27]: what do we Sqlite for here? isn't this specific to the tests?

File: timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+                };
+                
+                memoryContext.Add(group2);
+                memoryContext.Add(group2User3);
+                memoryContext.Add(group2Users4);
+                
+                await memoryContext.SaveChangesAsync();
+            }
+
+            {
```
Comment (daberni) [inline timrlink.net.CLI.Test/GroupUsersExportDatabaseActionTest.cs:187]: scopes? I don't really get the idea of this...

File: timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                var databaseGroup = context.Group
+                    .FirstOrDefault(g => g.ExternalId == group.externalId) ?? new Group();
+                
+                databaseGroup.Description = group.description;
+                databaseGroup.ExternalId = group.externalId;
+                databaseGroup.Name = group.name;
+                databaseGroup.ParentalExternalId = group.parentExternalId;
+
+                context.Update(databaseGroup);
```
Comment (daberni) [inline timrlink.net.CLI/Actions/GroupUsersDatabaseExportAction.cs:44]: AddOrUpdate ?

**Review by daberni — APPROVED** [review]

**Pull request merged**